horizon: rewrite AccountShowAction and TransactionIndexAction(account) endpoints

howardtw opened the pull request:
This PR:
1. removes `AccountShowAction` struct completely
2. removes the entry point in `main_generated.go`. We can delete the whole `main_generated.go` when we adopt this pattern.
3. moves `actions_account.go` into actions `package` and renames to it `account.go` directly.
4. introduces various http handlers to replace `base.go`. We can delete the whole `base.go` when we adopt this pattern. We can also delete `actions/responder.go` completely.
5. simply returns an error directly everywhere in the function when `err` is not `nil`. I found a place where we forgot to check the `base.Err` again in this PR. Please see comments in one of the files.  
6. removes the usage of `base` struct and `action` struct. We can remove the two structs as well.
7. doesn't use `App` (including the one in the context) in the `actions` package.

We should avoid putting the whole `App` in the context. A general rule of thumb is only putting things that's scoped to the request in the context. We can remove the `(*app) middleware` when we adopt this pattern.

tomquisel commented on services/horizon/internal/action.go:
```diff
@@ -152,3 +163,176 @@ func (action *Action) FullURL() *url.URL {
+type streamFunc func(context.Context, *sse.Stream) error
+type singleObjectStreamFunc func(context.Context) (sse.Event, error)
+
+func (a *App) streamableEndpointHandler(jfn jsonResponderFunc, sfn streamFunc, sosfn singleObjectStreamFunc) http.HandlerFunc {
```
I'd like to see a little more documentation on these functions, particularly on the arguments.

tomquisel commented on services/horizon/internal/action.go:
```diff
@@ -152,3 +163,176 @@ func (action *Action) FullURL() *url.URL {
+// Note that we cannot put this handler in the middleware stack because of
+// Chi's routing mechanism. A request will have to reach the end of the route
+// in order to have a valid route pattern in Chi.
+func accountIdHandler(h http.Handler) http.HandlerFunc {
```
The file seems like it has very general functions in it, but this `accountIdHandler` & friends functions seem pretty specific. Maybe break them out into another file?

howardtw replied:
Hmm I wanted to put all the functions associated with `actions` here. In other words, functions above `streamableEndpointHandler` will eventually be removed from this file because they are not really related to action. They exists because we embedded `App` in the action.

howardtw replied:
I tried to put `accountIdHandler` and its helper function in `middleware.go` but I realized I can't put it in the middleware stack. Let me know if you have a better idea for where these should belong. This is just semi-related to action.

tomquisel replied:
Gotcha, it's unfortunate they can't go in `middleware.go`. I'm not sure what to suggest, but I feel like there must be a better way. It's not a big deal, but worth a little more thought.

It seems like you have three different kinds of functions & methods:
1. methods of `App` like `streamHandler` and `streamableEndpointHandler` that provide general purpose handling of HTTP requests
2. functions like `accountIdHandler` that transform a handler, adding extra information from the request path to the HTTP request's context.
3. methods of `App` like `getAccountInfo`, which implement distinct endpoints.

Generally, I'd think that each of those 3 kinds of methods / functions would go in different files, since they don't depend on one another and are at different levels of abstraction (1. is general, 2. and 3. are more specific). 

Also, thinking about `accountIdHandler`, it's unfortunate to be doing our own string parsing of paths. I saw your comment about the issue with Chi's routing mechanism. There's no other way to take advantage of Chi's path parsing?

howardtw replied:
How about this?
1. `handlers.go`
2. `account.go`
3. `actions.go`

> There's no other way to take advantage of Chi's path parsing

I don't want to change the logic of how we get the information from the URL in this PR. I personally would like to not depend on chi at all in the future.

tomquisel replied:
That sounds like a good split to me! Is `account.go` too specific, though? What if we add other url parsers, would each go in their own file, or maybe they should go together?

howardtw replied:
https://github.com/stellar/go/pull/894/commits/c5b7aec03d721fcfd4a9e3732a9281e4bfa8994b
^ I've put handlers into `handler.go` for now to keep it simple because they are all going to be used to wrap the endpoints.

tomquisel commented on services/horizon/internal/hchi/context.go:
```diff
@@ -37,3 +38,16 @@ func RequestID(ctx context.Context) string {
 	return val
 }
+
+func WithAccountID(ctx context.Context, accountID string) context.Context {
```
Docs on these please

tomquisel commented on support/render/hal/handler.go:
```diff
@@ -0,0 +1,23 @@
+
+		res, err := fn(ctx)
+		if err != nil {
+			problem.Render(ctx, w, err)
```
maybe you need a `return` here?

howardtw replied:
Ah not again! Good catch 🤦‍♂️

bartekn commented on services/horizon/internal/handler.go:
```diff
@@ -0,0 +1,205 @@
+		if parts[0] != "accounts" {
+			h.ServeHTTP(w, r)
+			return
+		}
```
I don't understand why do we need to do this. It' only used in `/account/{id}` path. Is it a second check? I don't think it's needed.

howardtw replied:
It makes sure that when it's used in other endpoints, it won't do anything. If you don't like it, I can remove it.

bartekn commented on services/horizon/internal/handler.go:
```diff
@@ -0,0 +1,205 @@
+		}
+
+		ctx := r.Context()
+		addr, err := getAddress(r, "account_id", true)
```
Isn't is just simpler to get param in `AccountInfo` function? I think that this intermediate function that checks params and add them to context complicates the code a lot without giving any direct advantages. Handling a single endpoint request is done in two places: here and in `actions` package. If you want to add/change/remove a param you need to edit both...

howardtw replied:
What's scoped in the request should be in the context.

bartekn replied:
I don't think this is what `context` was designed for. From [blog post about context](https://blog.golang.org/context):
> The set of goroutines working on a request typically needs access to request-specific values such as the **identity of the end user, authorization tokens, and the request's deadline**.

The perfect example is a logger with request ID set added in one of the middlewares. You can later get it from the context and all log entries will automatically have request ID set (making it super simple to filter log entries by request).

I really don't think that request values (like GET, POST or URL params) should be added to the request context because you can easily get them from `http.Request` or `chi` methods.

Here's an interesting [article about `context` in web servers](https://www.calhoun.io/pitfalls-of-context-values-and-how-to-avoid-or-mitigate-them/).

bartekn commented on services/horizon/internal/handler.go:
```diff
@@ -0,0 +1,205 @@
+
+		contentType := render.Negotiate(r)
+		switch contentType {
+		case render.MimeHal, render.MimeJSON:
```
I don't think it's the right solution. I was hoping for more separation between `App` and HTTP server function. Instead, it looks like `App` is even bigger now (this code was previously in a different package at least).

howardtw replied:
I don't think `App` is any bigger. It's actually smaller. We use the thing in the app in the same way, but we don't have to assign the thing in the `App` to `Base` anymore.

bartekn replied:
You added 4 new methods on `App`. How is it not bigger? 😉

howardtw replied:
How does adding methods on App make App bigger? Please also see my comments here: https://github.com/stellar/go/pull/894#issuecomment-467960604

bartekn replied:
It simply adds some http-specific methods to `App`s responsibilities that were previously not there. That's why I'm saying you make `App` or it's responsibilities bigger. It's against separation of concerns principle. It makes it hard to test: ex. you can't unit test the action code without using `App`.

I think we should go in the opposite direction. `App` should really just initilize other components and make them work together.

howardtw commented:
> What we need is a separation between `App` and HTTP handlers (actions).

I agree with you @bartekn . We treat the `App` as our service, and we probably need another `Api` inside the `App` handling http requests. However, we should do it incrementally. This PR removes the need for both `Base` and `Action`. I think it's a good first step moving toward the better solution.

bartekn commented:
Sorry, but I don't feel we should merge this in the current form. Detailed explanations are in the comments above but to sum it up:
* Adding more responsibilities to `App` instead of the opposite: https://github.com/stellar/go/pull/894#discussion_r260901508
* Code will become harder to maintain (https://github.com/stellar/go/pull/894#discussion_r260852279) and test (https://github.com/stellar/go/pull/894#discussion_r260901508).
* I'm quite sure that the way `context` package is used in this PR is incorrect: https://github.com/stellar/go/pull/894#discussion_r260898344.

Let's discuss this next week when you're back.

howardtw commented:
I’m not sure if I can agree with you:
> Adding more responsibilities to App instead of the opposite.

I have mentioned that this is the first step of taking the App out of Actions. In the future, I see an `Api ` struct or `Web` struct in the `App` to handle http requests.

> Code will become harder to maintain

I’m not sure about this.

> I'm quite sure that the way context package is used in this PR is incorrect.

I think we both agree putting the whole App in the context as what we do now is incorrect. I think whether we should put the account address in the context is debatable as one can query others’ account . But the fact that an account endpoint has to parse the URL doesn’t feel right to me. It should already know who this is as you pointed out, the identify of the end user.

cc @tomquisel

bartekn commented:
> I have mentioned that this is the first step of taking the App out of Actions. In the future, I see an `Api ` struct or `Web` struct in the `App` to handle http requests.

The problem is that we're not in the future yet. We can merge this PR into a new experimental branch if you want (not `master`).

> > Code will become harder to maintain
>
> I’m not sure about this.

Any arguments supporting this?

> But the fact that an account endpoint has to parse the URL doesn’t feel right to me.

Yes, this should be a function in the internal `http`/`web` package. `actions` shouldn't do anything related to parsing requests. Separation of concerns.

> It should already know who this is as you pointed out, the identify of the end user.

No, this is not what I (actually Golang blog) said. Account ID is simply a parameter of the query. It's not the identity of the end user. The identity of a user would be, for example, email or user ID taken from session cookie. Such thing is a good candidate for being in a request context.

howardtw commented:
> The problem is that we're not in the future yet. We can merge this PR into a new experimental branch if you want (not `master`).

We work steps by steps toward a better future. For now, breaking down the path into PRs and moving along is what we can do. The fact that this PR removes the dependency on `Base` and change only one single points is a good little step in my opinion.

> Code responsible for handling a single action is in 2 places and test you need `App` to test action

What are the 2 places you are referring to? Like I said, later on we will have another struct called `Api` or `Web` in `App`, and we will just have to get a struct of that to test, not the whole `App`. Please bear in mind that this PR doesn't change how we test the endpoint as the test remains the same and they are passing.

> Account ID is simply a parameter of the query. It's not the identity of the end user.

We can definitely move the parsing logic out. I'm not gonna let this block us. After you reading the Golang blog, I'm pretty sure that you know that we shouldn't put the App in the context.

bartekn commented:
> We work steps by steps toward a better future. For now, breaking down the path into PRs and moving along is what we can do. The fact that this PR removes the dependency on `Base` and change only one single points is a good little step in my opinion.

I'm not against splitting this into many PRs. The problem is that it's solving one problem but introducing some new issues. Let's improve the code here first. I don't think we should rush important architecture decisions.

> What are the 2 places you are referring to?

[Here](https://github.com/stellar/go/pull/894/files#diff-8daceb6de5266af51ae92ffe4b715719R15) and [here](https://github.com/stellar/go/pull/894/files#diff-b8d16027eadf970091bce20ac2f34450R169).

> After you reading the Golang blog, I'm pretty sure that you know that we shouldn't put the App in the context.

Yes, but it's not what we're discussing here. The whole `context` convo was about putting request params in context. You solve "`App` in context" problem and introduce "request params in context" problem.

---

It's pretty late here so if you don't mind I'll spend some time later this week (you're OOO anyway) thinking about changes required before merging this.

howardtw commented:
@bartekn sounds like if I move the account id handling part into the `AccountInfo` function, everything will be resolved.

bartekn commented:
> My take is that if this is an improvement, why do we have to wait to get this in? Especially, this only affects a single endpoint. We know this is better than the current architecture after moving the account id part into the function.

It improves some things but makes other things difficult or unclear (already explained that in comments above). I really don't want to rush architectural changes like this, even if it's a single endpoint change. I prefer to have a bigger discussion where we can find pros and cons of proposed solutions and pick the best one. I don't think experimenting on master (outside `exp` package) is a good idea.

For example, when I started thinking about the design (some ideas below) I realised that we probably shouldn't remove actions structs (like `AccountShowAction`). It's fine for non-streaming endpoint but for streaming endpoints it would be great to have a place to store some data between calls to `SSE()` (like request params so we don't have to parse them every time or the current cursor so we don't need to do slice cutting like now).

---

Let's start with the requirements for the new design:
* Separation of concerns: `App` should initilize other components and make them work together, it shouldn't know how http server works, how actions work etc.
* Move actions to a separate package.
* Easier to test. If we want to test action, we don't shouldn't have to create `App` instance, only mock object used directly by actions (ex. db session objects).
* All methods return errors, no internal `Err` fields.
* And of course: no `App` in `context` :).

Here's a rough idea what we can do:
* `http` package where http server is created, router is initialized, middlewares added etc. Exposes a public function used to add `routes -> handlers` mapping (routes and handlers will be provided by `App`). We move code responsible for action rendering (basically `actions.Base.Execute`) there too, including interfaces. (Can be probably merged with existing `httpx` package.)
* `actions` package with actions code. I think that in the first iteration we should probably just remove:
  * `Err`: all internal methods return errors,
  *  `appCtx`: `http` package should use `Prepare` function on actions to provide all necessary objects, like DB handler etc.
* `App` is used to get actions and pass them to `http` (or maybe, since actions shouldn't be needed by `App` anywhere, `http` should just import `actions` package itself?).

With something like this:
* We make `App` smaller - all actions and web related things are moved to corresponding packages.
* When `http` provides all necessary objects to actions there's no need to keep `App` in the request context.
* We can unit test actions by providing it with mock objects (ex. mock DB object that returns some test rows).

Let me know what you think.

howardtw commented:
> It improves some things but makes other things difficult or unclear.

I'm not sure what would become difficult and unclear. If you think it makes testing difficult, I would say we don't have those tests yet as I didn't modify any tests in this PR and all the current tests are passing. If you insist that we should include the unit (struct) in `App` handling http requests in this PR, I can also do it. I just prefer not to as it's going to be involved much more changes and makes this PR too big.

> I prefer to have a bigger discussion where we can find pros and cons of proposed solutions and pick the best one.

while it sounds reasonable, I politely disagree this approach under the presumption this PR already satisfies (or will) all the requirements you listed above. One of the reasons that we should make all the PRs small is that people can review an improvement very easily.

> I don't think experimenting on master (outside exp package) is a good idea.

Making PRs small makes master a good candidate for changes. I don't see why we can't do it on master if we do it incrementally.

> I realised that we probably shouldn't remove actions structs (like AccountShowAction). It's fine for non-streaming endpoint but for streaming endpoints it would be great to have a place to store some data between calls to SSE() (like request params so we don't have to parse them every time or the current cursor so we don't need to do slice cutting like now).

I believe this PR doesn't change how to store data between the calls to SSE(). We should discuss further improvement outside of this PR.

> Let's start with the requirements for the new design:

This is a good list of requirements 👍 

> Separation of concerns: App should initialize other components and make them work together, it shouldn't know how http server works, how actions work etc.

This will be handled in the following PR. I don't feel like removing things and adding things in the same PR as it will make this PR too big. However, given that we have been discussing this for a while, I won't insist on not doing it this PR.

> Move actions to a separate package.

Done in this PR.

> Easier to test. If we want to test action, we don't shouldn't have to create App instance, only mock object used directly by actions (ex. db session objects).

This falls in the same requirement as the first one.

> All methods return errors, no internal Err fields.

Done in this PR.

> And of course: no App in context :).

Done in this PR.

> http package where http server is created, router is initialized, middlewares added etc. Exposes a public function used to add routes -> handlers mapping (routes and handlers will be provided by App). We move code responsible for action rendering (basically actions.Base.Execute) there too, including interfaces. (Can be probably merged with existing httpx package.)

I'm supportive to have a `http` packages, but I think initializing the routes, the server, and the middleware can just be done in the top level of the `horizon` package. action rendering might be a good candidate to put in the package. So far, I put the action rendering code in `support/render/hal`. Maybe we can put `render/hal` under the http package. Again, adding a new package to consolidate the logic about how we handle http requests can be a PR that's orthogonal to this one.

> actions package with actions code. I think that in the first iteration we should probably just remove:
Err: all internal methods return errors

Done in this PR.

> appCtx: http package should use Prepare function on actions to provide all necessary objects, like DB handler etc.

I think we don't need a `Prepare` function nor a `action` struct anymore as we are going to introduce units in `App` and handle requests. What might be good is to have two different units in `App`, one for handling json requests, the other for handling streaming requests.

> App is used to get actions and pass them to http.

Agree. But I think since the logic passing the request from `App` to the unit that handles either json or streaming won't be in the `http` package, itself, it would just be `horizon` package importing `action` package.

---
At this point, I feel like we should probably loop @tomquisel in to unblock us. I think we are both working toward a better code structure, but the approach we take is a little different. It would be great to have a third person in our conversation to get things going.

bartekn commented:
OK, what do you think about short discussion tomorrow after standup? Will be faster than writing comments 😉.

howardtw commented:
I'm on board 😃

howardtw commented:
Hi guys @tomquisel @bartekn 👋 

I've brought this PR to a state where it's ready to be reviewed. I've found so many missing err checks and repetitive calls to the same function when I looked at the code where how we get page params. I've organized all the new url helper functions in `urlparams.go` and streamline the way we load the necessary params in a `transactionHandler`. Of course, return immediately when we encounter an error wrapped with a detailed message.

I think `web` is the thing we are looking for to handle all the api traffic, so I added a few field that's needed and unexported the struct since it's not used by other packages. I've also added a couple tests in `action_test.go`, `actions/account_test.go`, and `actions/transaction_test.go`. Of course there are more tests and function comments to be added to this PR.

I think this one is very close to completion, and I'll start to focus on community PRs and issues. As I mentioned, expanding this PR to another endpoint will make this PR very big and this harder to review. Please take your time and give it a review. I want to close this out as soon as I can to focus on our OKR. Not to say that this is not important. By making it go idiomatic, we will be able to have a better understanding of the system with an improved error handling mechanism and make our system more robust.

howardtw commented on services/horizon/internal/action.go:
```diff
@@ -152,3 +169,42 @@ func (action *Action) FullURL() *url.URL {
+func (w *web) loadAccountEvent(ctx context.Context, param interface{}) (sse.Event, error) {
+	res, err := w.getAccountInfo(ctx, param)
+	return sse.Event{Data: res}, err
+}
```
We only have a few single object streamers and given this kinda function is minimal, I think it wouldn't be an issue.

howardtw commented on services/horizon/internal/actions/transaction.go:
```diff
@@ -0,0 +1,106 @@
+	var records []history.Transaction
+
+	txs := hq.Transactions()
+	txs.ForAccount(addr)
```
I already stated that here:
> I would expect that a lot of the new code in this PR can be reusable by other endpoints, especially for transaction-by-ledger_id and the verification on how we get the params.

One approach is to pass in a new type with possible values to be`byAccount` and `byLedger`. However, this is not in the scope of this PR.

howardtw commented on services/horizon/internal/actions/transaction.go:
```diff
@@ -0,0 +1,106 @@
+	}
+
+	s.SetLimit(int(pq.Limit))
+	records := allRecords[s.SentCount():]
```
This PR is not aiming to resolve this issue actually.

bartekn replied:
AFAIR, this was meant to be experimental PR that demonstrates a new architecture of actions in Horizon (#869). Keeping state of a stream was something we discussed during a call with @tomquisel as something that would improve streaming code (in action handlers).

howardtw replied:
Sorry it might be my bad. I somehow thought that when you said the state of a stream, you meant the filters used during streaming in `TransactionIndexAction`. For this issue, we can potentially move the `stream.send` logic out from the function and put it in the `streamHandler` and track the number of sent event with a local variable in the `streamHandler` in future PRs.

howardtw commented on services/horizon/internal/handler.go:
```diff
@@ -0,0 +1,311 @@
+// between this function and streamFunc is that this one only loads an event. The
+// server will not send an event out if the current event is same as the last one.
+// Please see the implementation of streamHandler for more details.
+type singleObjectStreamFunc func(context.Context, interface{}) (sse.Event, error)
```
How is it too many? Can we elaborate what the issue is with the old transaction builder? I can write a better comment here to explain what to pass in if you think it's hard to understand. Basically developers can decide whatever type they want to load the object.

bartekn replied:
The problem with `interface{}` is that it doesn't do type checking on compilation time so you can pass/return invalid type and you'll find out when the code is running. It's also not clear just looking at the function what it wants. For example:
```go
func (w *web) getAccountInfo(ctx context.Context, param interface{}) (interface{}, error) {
```
Just looking at this function I can't tell if param is `string`, some kind of request `struct`, maybe `url.Values` or even `http.Request`. Return value can be `struct`, `[]byte`. You see what I mean? I'd say that `interface{}` gives a great flexibility but at the cost of readability of the code.

I'm not Go guru so I can be wrong but I mentioned `build` package example. See a function that creates account_merge operation: [`func AccountMerge(muts ...interface{}) (result AccountMergeBuilder)`](https://godoc.org/github.com/stellar/go/build#AccountMerge). Developers had problems understanding what they need to pass to these functions. What is more, we at SDF quickly also forgot how to use this thing too. The only way to use these code reliably was to copy examples from godoc.

howardtw replied:
I see what you mean. For this we might be able to improve it with `reflection` in the future, but I don't want to get that far in this PR. For now we can have a good comment for the function and make sure we don't pass the interface down to the actual function in the `actions` package.

tomquisel replied:
The closure suggestion above should take care of the `interface{}` here, at least in the argument.

howardtw commented on services/horizon/internal/handler.go:
```diff
@@ -0,0 +1,311 @@
+		return nil, errors.Wrap(err, "getting account address")
+	}
+
+	lid, err := getInt32ParamFromURL(r, "ledger_id")
```
I'm not sure if every single endpoint will use the same set of params at this point. Given the scope of this PR, I think it makes sense to do it like this. We can always reuse the same code and give the functions a more general name.

For example, `transactionHandler` might end up the same as `operationHandler`, and `loadTransactionParams` might end up the same as `loadOperationParams`. Needless to say, we already know that those `ByAccount` can be removed when `ByLedger` is introduced.

bartekn replied:
> And I would argue that one has to implement the following functions under the current structure

There are actually 3 main differences:
* The function you listed are in a single file (except one - see last point), within a single package. Again, I think that structure you present here would make sense if params parsing and passing to the final action code was done automatically (something like request structs - `TransactionParams` in your code). Currently there's a lot of boilerplate code required to actually get and pass these params. This could be done by a single line in the action code.
* These 6 functions actually handle 3 endpoints: Transaction for Account, Ledger and All transactions. I see your comment about merging all `TransactionPageBy*` methods in one now but it'd be great to see how a function like this would look like.
* Number 1 on your list (`TransactionIndexAction.Handle`) is generated automatically by `go generate` whereas you need to create functions manually.

howardtw replied:
> The function you listed are in a single file (except one - see last point), within a single package. 

One of the main goals of this PR is to separate the packages. I believe the actions_* files should be moved into the `actions` package. (Maybe we should rename to package to `query`?) 

>  I think that structure you present here would make sense if params parsing and passing to the final action code was done automatically (something like request structs - TransactionParams in your code). Currently there's a lot of boilerplate code required to actually get and pass these params. This could be done by a single line in the action code.

I will definitely reuse the same code if possible when this gets expanded to other endpoints. I can see that we have a struct called `QueryParams` in the future, and all the endpoints share the same struct. If this is true, then the `interface{}` readability problem would be resolved too :)

tomquisel replied:
I'm not really a fan of creating a mega-struct that contains the params for all endpoints. This creates an interdependency between endpoints. I honestly can't think of a way to do this better than what you have here given our code. @bartekn or @howardtw can you any better approach? A generic param getter like you're suggesting, Bartek, will still need to do some conversion of those parameters, and seems like you'll just be piling all these separate parameter-parsing functions into one big one.

bartekn commented:
OK, thanks for explanations. I added some comments. Would be great if we could think about solution to: https://github.com/stellar/go/pull/894#discussion_r267111991 and https://github.com/stellar/go/pull/894#discussion_r267119807 but I guess it can be done in another PR. Even though I still think some improvements can be made, maybe it's fine to merge it and try it on `master` in day to day development.

tomquisel commented on services/horizon/internal/action.go:
```diff
@@ -152,3 +154,42 @@ func (action *Action) FullURL() *url.URL {
+
+// getAccountInfo returns the information about an account based on the provided param.
+// The expected param here is the account id, which has a string type.
+func (w *web) getAccountInfo(ctx context.Context, param interface{}) (interface{}, error) {
```
My suggestion for avoiding the `interface{}` argument here and in `loadAccountEvent` is to change what `web.accountHandler` does. Instead, it can explicitly take a function with signature `(ctx context.Context, addr string) (interface{}, error)` (and do the equivalent for the stream function), and then modify `web.accountHandler` to create a closure wrapping `getAccountInfo` and calling it with `addr`. Like so: `we.streamableEndpointHandler(func(ctx context.Context) { jfn(ctx, addr) }, ...)`.

tomquisel commented on services/horizon/internal/action.go:
```diff
@@ -152,3 +154,42 @@ func (action *Action) FullURL() *url.URL {
+
+// loadAccountEvent loads the information about an account based on the provided param into a server-sent event.
+// The expected param here is the account id, which has a string type.
+func (w *web) loadAccountEvent(ctx context.Context, param interface{}) (sse.Event, error) {
```
Is there any systematic reason for prefixing functions with `load` vs `get`? I usually use `load` when it's going to go to a database or some external source and `get` when it's a simple local computation. I might update it here and several other places to that convention unless you've got another convention already.

tomquisel commented on services/horizon/internal/handler.go:
```diff
@@ -0,0 +1,326 @@
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		err := errorIfHistoryIsStale(we.isHistoryStale())
```
It seems like this kind of validation of arguments should happen in the `action` handler itself (since it applies regardless of whether you call the action from HTTP or something else like GraphQL). Here it's intimately tied to extracting parameters from a URL.

howardtw replied:
That's a good point. For now I think it might be good to move the check into `w.coreSession` and `w.horizonSession` so that whoever wants to grab a database handle we run the check automatically.
